console/unicode_width_test.ts

```diff
@@ -0,0 +1,38 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+import { unicodeWidth } from "./unicode_width.ts";
+import { assertEquals } from "../testing/asserts.ts";
+
+Deno.test("unicodeWidth", async (t) => {
+  await t.step("ASCII", () => {
+    const lorem =
+      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
+
+    assertEquals(unicodeWidth(lorem), lorem.length);
+  });
+
+  await t.step("CJK", () => {
+    const qianZiWen =
+      "天地玄黃宇宙洪荒日月盈昃辰宿列張寒來暑往秋收冬藏閏餘成歲律呂調陽雲騰致雨露結爲霜金生麗水玉出崑岡劍號巨闕珠稱夜光果珍李柰菜重芥薑海鹹河淡鱗潛羽翔龍師火帝鳥官人皇始制文字乃服衣裳推位讓國有虞陶唐弔民伐罪周發殷湯坐朝問道垂拱平章愛育黎首臣伏戎羌遐邇壹體率賓歸王鳴鳳在樹白駒食場化被草木賴及萬方蓋此身髮四大五常恭惟鞠養豈敢毀傷女慕貞絜男效才良知過必改得能莫忘罔談彼短靡恃己長信使可覆器欲難量墨悲絲淬詩讚羔羊";
+
+    assertEquals(unicodeWidth(qianZiWen), qianZiWen.length * 2);
+  });
+
+  await t.step("Unicode normalization", async (t) => {
+    const str = "á";
+
+    await t.step("NFC", () => {
+      const nfc = str.normalize("NFC");
+
+      assertEquals(nfc.length, 1);
+      assertEquals(unicodeWidth(nfc), 1);
+    });
+
+    await t.step("NFD", () => {
+      const nfd = str.normalize("NFD");
+
+      assertEquals(nfd.length, 2);
+      assertEquals(unicodeWidth(nfd), 1);
+    });
+  });
```

Comment:
```suggestion
  });

  await t.step("Emoji", () => {
    assertEquals(unicodeWidth("👩"), 2); // Woman
    assertEquals(unicodeWidth("🔬"), 2); // Microscope
    assertEquals(unicodeWidth("👩‍🔬"), 4); // Woman Scientist
  });
```

I think additional tests could be useful — they serve as documentation of behavior. That's one from the upstream [readme](https://github.com/unicode-rs/unicode-width/blob/v0.1.17/README.md?plain=1#L27-L40).

Reply:
Good point. I don't think this is an intentional behavior. What we want here is to calculate the visual width of characters in terminals.

Note: `npm:string-width` returns 2 for 👩‍🔬

```js
import stringWidth from 'npm:string-width';
console.log(stringWidth("👩‍🔬"));
```

Reply:
@kt3k

> I don't think this is an intentional behavior.

It is intentional. The term _width_ **doesn't** mean "visual column width" in either the upstream crate or the unicode technical report from which it is derived.

> What we want here is to calculate the visual width of characters in terminals.

Are you sure? [I don't think that's going to be possible without much deeper knowledge about the terminal environment](https://stackoverflow.com/q/71228251/438273) — and I'm not sure whether the required info can be reliably obtained (especially cross-platform in many terminals).... at the least, it will be a difficult task.

> Note: `npm:string-width` returns 2 for 👩‍🔬

`👩‍🔬` occupies `1` column in my current terminal environment.

Reply:
> It is intentional. The term width doesn't mean "visual column width" in either the upstream crate or the unicode technical report from which it is derived.

That's true for the upstream crate, but I don't think it's true for this module.

This PR starts from this issue https://github.com/denoland/deno/issues/18477 (and https://github.com/denoland/deno_std/issues/3296 )

We want to know how much characters can fit in `Deno.consoleSize()`

Reply:
> 👩‍🔬 occupies 1 column in my current terminal environment.

What OS and terminal do you use? How does the script `console.log("👩‍🔬".repeat(Deno.consoleSize().columns))` render in your terminal?

Reply:
@kt3k

> This PR starts from this issue [denoland/deno#18477](https://github.com/denoland/deno/issues/18477) (and #3296 )
> 
> We want to know how much characters can fit in `Deno.consoleSize()`

Understood — thanks for the linked information.

> What OS and terminal do you use?

```
% sw_vers
ProductName:		macOS
ProductVersion:		13.3.1
BuildVersion:		22E261
```

```
% system_profiler -json SPApplicationsDataType | gojq '.SPApplicationsDataType[] | select(._name=="Terminal")'
{
  "_name": "Terminal",
  "arch_kind": "arch_arm_i64",
  "lastModified": "2023-04-01T16:46:50Z",
  "obtained_from": "apple",
  "path": "/System/Applications/Utilities/Terminal.app",
  "signed_by": [
    "Software Signing",
    "Apple Code Signing Certification Authority",
    "Apple Root CA"
  ],
  "version": "2.13"
}
```

> How does the script `console.log("👩‍🔬".repeat(Deno.consoleSize().columns))` render in your terminal?

```
% deno eval 'console.log("👩‍🔬".repeat(Deno.consoleSize().columns))'
👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬👩‍🔬
```

I'm also using a customized profile in Terminal.

Reply:
Thanks for sharing!

The way Terminal.app handles it looks strange (I was also able to reproduce it.)

On the other hand, iTerm2 seems handling the combined emoji in a better way.

In iTerm2
<img width="1433" alt="Screen Shot 2023-04-15 at 18 30 33" src="https://user-images.githubusercontent.com/613956/232205620-5582abea-394f-48ad-9ead-78b946b66492.png">

In the above (iterm2) case returning the width (or columns?) 2 feels sensible to me.

Anyway the combined emoji cases look complicated and even having incompatibilities among terminal software. I'd suggest we'd leave this as is and iterate on them later.

Reply:
> I'd suggest we'd leave this as is and iterate on them later.

Does that mean including the [suggested test](https://github.com/denoland/deno_std/pull/3297#discussion_r1166289430)?

Reply:
Let's add them with comments referencing to this PR

Reply:
> Let's add them with comments referencing to this PR

[Comment added](https://github.com/denoland/deno_std/pull/3297#discussion_r1166289430)